### PR TITLE
Forward tensor.insert_slice coming from in_parallel lowering to flow.…

### PR DESCRIPTION
**Pull request by nicolasvasilache**
…dispatch.tensor_store

This pattern is currently necessary for correctness, it accounts for the fact that
InParallel is distributed across multiple workgroups when lowering to HAL
but the current implementation connects it to a sequential tensor.insert_slice
and only later to flow.dispatch.tensor_store.

In the future, all the rewrites in this file this should be done as part of InParallel
-> HAL rewrite. But because of dialect dependencies and layering, we have
some phase ordering that prevents it atm.

A similar layering issue prevents bufferization to be controlled by the transform dialect atm.
The effects of the insert_slice -> tensor_store forwarding are best observed before bufferization
removes the tensor_store op.
To allow more modular testing and separation of concerns, a temporary test-only flag is added to
disable bufferization and exhibit the proper forwarding behavior.

Finally, a previously failing integration test is now correct.

### CODE REVIEW

File: iree/compiler/Codegen/Common/SetNumWorkgroupsFromLinalgExtPass.cpp [hunk 1]
```diff
@@ -60,6 +62,70 @@ class OneToOneRewritePattern : public OpRewritePattern<SourceOp> {
...
+/// Forward LinalgExt::InParallel -> Tensor::InsertSlice -> Flow::TensorStore.
+/// This pattern is necessary for correctness, it accounts for the fact that
+/// InParallel is distributed across multiple workgroups when lowering to HAL
+/// but it then connects to a sequential tensor.insert_slice and then to
+/// flow.dispatch.tensor_store.
+///
+// TODO: All the rewrites in this file this should be done as part of InParallel
+// -> HAL rewrite. But because of dialect dependencies and layering, we have
+// some phase ordering that prevents it atm.
+class ForwardInParallelResultToFlow
```
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/Common/SetNumWorkgroupsFromLinalgExtPass.cpp:74]: THis is the exact same pattern as https://github.com/google/iree/blob/879f07684af55ca7948b64cc00db460167fd263c/iree/compiler/Codegen/Common/BufferizeCopyOnlyDispatchesPass.cpp#L162 AFAICS. 

THose patterns are going to move to Flow as canonicalizations anyway.  Is there more than `tensor.insert_slice` -> `flow.dispatch.tensor.store` forwarding happening here?

**Review by MaheshRavishankar — APPROVED** [review]

File: iree/compiler/Codegen/Common/SetNumWorkgroupsFromLinalgExtPass.cpp (hunk 1 above)
Comment (nicolasvasilache) [reply iree/compiler/Codegen/Common/SetNumWorkgroupsFromLinalgExtPass.cpp:74]: Nope, just didn't know better as to look them up..
As I mentioned offline, I may as well adopt your patterns directly.

I was pondering moving them to FlowOps.cpp but that would create a HAL -> Flow dependence that is not there yet.

If you point me to a place where to move that code and expose it I can do that before landing?
I would only create the `populateFoldingXXXPatterns` and move code, not activate as folding/canonicalization in this patch.

Comment (MaheshRavishankar) [reply iree/compiler/Codegen/Common/SetNumWorkgroupsFromLinalgExtPass.cpp:74]: No need to move it to `Flow`. You can just add the `populateXXX` method to `iree/compiler/Codegen/Utils.h`/`Utils.cpp`. I can move it from there.

Comment (nicolasvasilache) [reply iree/compiler/Codegen/Common/SetNumWorkgroupsFromLinalgExtPass.cpp:75]: Upon further inspection, the application logic is not exactly the same, there is still a small impedance gap between inparallel and dispatch ops that will be ironed out over time. 

In the meantime I was able to refactor the core logic a bit (and we should also upstream it).

Comment (MaheshRavishankar) [reply iree/compiler/Codegen/Common/SetNumWorkgroupsFromLinalgExtPass.cpp:74]: Hmmm, I am not sure what the difference is. If you could file a bug describing the mismatch, that would be great.

**Pull request merged**